### ISSUE TITLE
Replacement can't happen at end of block without causing repeated replacements

### ISSUE THREAD

**Issue by terminalmage** [AskMeAgain/logseq-regex-linker#2]
## Problem

The example configuration in this project is as follows:

```
(PROJ-[0-9]*)\s::::[$1](http://cloud-jira-whatever.com/id/$1)
```

There is implied whitespace at the end of the match. This means that you cannot replace text at the end of the ~line~ block, because Logseq trims trailing whitespace when you leave a block.

I tried removing the `\s`, but when I do so, it matches on the link text (for example `PROJ-123`) from both of the `$1` replacements, resulting in multiple cascading replacements, which are repeated each time I leave and re-enter the block.

## Proposed Solution

To fix this, you could do the following:

1. Imagine a configuration with a single regex map with no whitespace at the end of the match:
   ```
   (PROJ-[0-9]*)::::[$1](http://cloud-jira-whatever.com/id/$1)
   ```
2. For each regex map in the configuration, do the following:
   1. Get all unique matches for the regex. Imagine the following block:
      ```
      TODO Follow up on both [PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123) and [PROJ-456](http://cloud-jira-whatever.com/id/PROJ-456)
      ```
      For this block, the unique regex matches would be `PROJ-123` and `PROJ-456`
   2. For each unique match, do the following:
      1. Generate the replacement for this match. For example, when processing `PROJ-123` the replacement would be `[PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123)`.
      2. Split the block text on the replacement. I'm not great at JavaScript but the equivalent Python would be something like:
         ```python
         sub_blocks = block.split(replacement)
         ```
      3. `sub_blocks` would be an array of substrings not containing the replacement text. For each substring, replace each occurrence of the unique match with the replacement. Note that this should be done with substring replacement, not regex replacement.
      4. Join `sub_blocks` by the replacement. Again, I'm not great at JavaScript but the equivalent Python would be:
         ```python
         block = replacement.join(sub_blocks)
         ```

Using this method, you can avoid making cascading replacements. It also allows the plugin to react to new cases of a match found when a block is edited. Imagine this example block:

```
[PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123) needs review
```

This is an contrived example, but suppose you wanted to add more text to this block, including another reference to the Jira ticket ID. Your edited block would then contain something like:

```
[PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123) needs review something something PROJ-123
```

Using the above method, we would avoid making cascading replacements on the markdown hyperlink at the beginning of the block, and would only replace the new `PROJ-123` reference, resulting in:

```
[PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123) needs review something something [PROJ-123](http://cloud-jira-whatever.com/id/PROJ-123)
```

**Comment by AskMeAgain** [comment]
Hi,

I noticed this issue aswell and i have a workaround for you

> Using this method, you can avoid making cascading replacements. It also allows the plugin to react to new cases of a match found when a block is edited.

The overall issue is that once you touch a line again, the replacement is running again. So you need to make sure that the original value is different to whatever it is getting replaced with (that is why you have the whitespace in the regex, as the url in the end doesnt have the whitspace). 

Your proposed solution will sadly not work as the cascading replacement is happening whenever you retouch a line (as this will trigger the plugin)

**Workaround: Simple add a short description to the end of the line after the jiraticket, so the whitespace is not getting removed.**

> "PROJ-1234 delete table column abc"

Or you make the regex match "PROJ_" as a marker. 

Or maybe a bit crazy but you can make a regex which matches PROJ-1234`NOT [ or )` then it wouldnt act on itself

Did you try to match on linebreaks? Maybe that is also a solution.

Overall i dont really see how i can fix this issue as this would require state (storing in a file or DB) to understand where the replacement happened and where it didnt. That is quite tricky

**Comment by terminalmage** [comment]
> The overall issue is that once you touch a line again, the replacement is running again. So you need to make sure that the original value is different to whatever it is getting replaced with (that is why you have the whitespace in the regex, as the url in the end doesnt have the whitspace).
> 

No. Not with my solution.

> Your proposed solution will sadly not work as the cascading replacement is happening whenever you retouch a line (as this will trigger the plugin)
> 

Yes, it will work. My solution accounts for this by splitting on the expected replacement, thus ignoring already-applied regex replacements and focusing on the areas of the block that aren't part of an already-implemented replacement.

> Overall i dont really see how i can fix this issue as this would require state (storing in a file or DB) to understand where the replacement happened and where it didnt. That is quite tricky

I don't know where you're getting the idea that state would need to be saved. Perhaps you didn't read my solution closely enough?

I don't know JavaScript well enough but here it is in Python:

```python
#!/usr/bin/env python3

import re

# Note: Python regex replacement uses "\1" instead of "$1"
regex_maps: list = [
    r"(PROJ-[0-9]*)::::http://cloud-jira-whatever.com/id/\1"
]

def do_replacement(block: str) -> str:

    print(f"OLD: {block}")

    for regex_map in regex_maps:
        match_pat: str
        replace_pat: str
        try:
            match_pat, replace_pat = regex_map.split("::::")
        except ValueError:
            continue

        for unique_match in frozenset(re.findall(match_pat, block)):
            replacement: str = re.sub(match_pat, replace_pat, unique_match)
            sub_blocks: list = block.split(replacement)

            index: int
            sub_block: str
            for index, sub_block in enumerate(sub_blocks):
                sub_blocks[index] = sub_blocks[index].replace(unique_match, replacement)
            block = replacement.join(sub_blocks)

    print(f"NEW: {block}")
    print("----------------------------------------")

    return block


if __name__ == "__main__":

    contents: str = "TODO Follow up on both PROJ-123 and PROJ-456"

    # Run replacement once
    contents = do_replacement(contents)
    # Run it again....
    contents = do_replacement(contents)
    # And again....
    contents = do_replacement(contents)
    # And again....
    contents = do_replacement(contents)

    # Add some separator lines before proceeding
    print("")
    print("========================================")
    print("========================================")
    print("========================================")
    print("")

    # Now let's add another "PROJ-123" to the block's contents
    contents += " and assign PROJ-123 to Jonas"

    # Run the replacement again
    contents = do_replacement(contents)
    # And again....
    contents = do_replacement(contents)
    # And again....
    contents = do_replacement(contents)
```

And here is what you get when you run that code:

```
OLD: TODO Follow up on both PROJ-123 and PROJ-456
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
----------------------------------------
OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
----------------------------------------
OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
----------------------------------------
OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456
----------------------------------------

========================================
========================================
========================================

OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign PROJ-123 to Jonas
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign http://cloud-jira-whatever.com/id/PROJ-123 to Jonas
----------------------------------------
OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign http://cloud-jira-whatever.com/id/PROJ-123 to Jonas
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign http://cloud-jira-whatever.com/id/PROJ-123 to Jonas
----------------------------------------
OLD: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign http://cloud-jira-whatever.com/id/PROJ-123 to Jonas
NEW: TODO Follow up on both http://cloud-jira-whatever.com/id/PROJ-123 and http://cloud-jira-whatever.com/id/PROJ-456 and assign http://cloud-jira-whatever.com/id/PROJ-123 to Jonas
----------------------------------------
```

You could run that code a million times and it would not perform cascading replacements. This solution prevents the need for all of the odd workarounds you suggested.

Please let me know if you decide not to implement this fix. This plugin is useful enough to me that I will gladly fork and learn enough JavaScript to implement it myself, if that is what must be done.

EDIT: I just realized that in my POC code above, the replacement wasn't in the form of a markdown hyperlink, but the POC still shows that it will not make cascading replacements when run repeatedly, which is what is important.

**Comment by AskMeAgain** [comment]
Hi, 

indeed i wasnt understanding your solution 👍  I implemented your solution and pushed v2.0.0, you should be able to update the plugin now and try it out :)

There is one edgecase i found and that is when the line contains *only* the matching string, then splitting the line will result in an empty result set, but i fixed that by introducing whitespaces at the beginning and end of the line to make sure that whenever we split we get an array with entries. Logseq afterwards will remove the trim automatically so this is just an implementation detail.

**Comment by terminalmage** [comment]
> Hi,
> 
> indeed i wasnt understanding your solution 👍 I implemented your solution and pushed v2.0.0, you should be able to update the plugin now and try it out :)
> 

:raised_hands: 

> There is one edgecase i found and that is when the line contains _only_ the matching string, then splitting the line will result in an empty result set, but i fixed that by introducing whitespaces at the beginning and end of the line to make sure that whenever we split we get an array with entries. Logseq afterwards will remove the trim automatically so this is just an implementation detail.

Do you mean if the block's contents contain only `PROJ-123`? Or only the replacement text?
